backends/tc/ebpfCodeGen.cpp

```diff
@@ -408,13 +408,173 @@ void TCIngressPipelinePNA::emitLocalVariables(EBPF::CodeBuilder *builder) {
 }
 
 // =====================EBPFPnaParser=============================
+EBPFPnaParser::EBPFPnaParser(const EBPF::EBPFProgram *program, const IR::ParserBlock *block,
+                             const P4::TypeMap *typeMap)
+    : EBPFPsaParser(program, block, typeMap) {
+    visitor = new PnaStateTranslationVisitor(program->refMap, program->typeMap, this);
+}
+
 void EBPFPnaParser::emit(EBPF::CodeBuilder *builder) {
     builder->emitIndent();
     builder->blockStart();
     EBPF::EBPFParser::emit(builder);
     builder->blockEnd(true);
 }
 
+void PnaStateTranslationVisitor::compileExtractField(const IR::Expression *expr,
+                                                     const IR::StructField *field,
+                                                     unsigned alignment, EBPF::EBPFType *type) {
+    unsigned widthToExtract = dynamic_cast<EBPF::IHasWidth *>(type)->widthInBits();
+    auto program = state->parser->program;
+    cstring msgStr;
+    cstring fieldName = field->name.name;
+
+    bool checkIfMAC = false;
+    auto annolist = field->getAnnotations()->annotations;
+    for (auto anno : annolist) {
+        if (anno->name != ParseTCAnnotations::tcType) continue;
+        auto annoBody = anno->body;
+        for (auto annoVal : annoBody) {
+            if (annoVal->text == "macaddr") {
+                checkIfMAC = true;
+                break;
+            }
+        }
+    }
+
+    msgStr = Util::printf_format("Parser: extracting field %s", fieldName);
+    builder->target->emitTraceMessage(builder, msgStr.c_str());
+
+    if (widthToExtract <= 64) {
```

Comment:
We need to update the code to use memcpy for macaddr. Perhaps a lot of code will go away from this

Reply:
resolved